api/oneclick.py

```diff
@@ -0,0 +1,453 @@
+                patron_oneclick_id, item_oneclick_id)
+            return True
+
+        # should never happen
+        raise CirculationException("Unknown error %s/%s checking in %s.", patron.authorization_identifier, 
+            patron_oneclick_id, item_oneclick_id)
+
+
+    def checkout(self, patron, pin, licensepool, internal_format):
+        """
+        Associate an ebook or audio with a patron.
+
+        :param patron: a Patron object for the patron who wants to check out the book.
+        :param pin: The patron's password (not used).
+        :param licensepool: The Identifier of the book to be checked out is 
+        attached to this licensepool.
+        :param internal_format: Represents the patron's desired book format.  Ignored for now.
+
+        :return LoanInfo on success, None on failure
+        """
+        (patron_oneclick_id, item_oneclick_id) = self.validate_input(patron, licensepool)
+
+        resp_dict = self.circulate_item(patron_id=patron_oneclick_id, item_id=item_oneclick_id, return_item=False)
+
+        if not resp_dict or ('error_code' in resp_dict):
+            return None
+
+        self.log.debug("Patron %s/%s checked out item %s with transaction id %s.", patron.authorization_identifier, 
+            patron_oneclick_id, item_oneclick_id, resp_dict['transactionId'])
+
+        today = datetime.datetime.now()
+        expires = today + self.EXPIRATION_DEFAULT
+
+        # Create the loan info. We don't know the expiration 
+        loan = LoanInfo(
+            identifier_type=licensepool.identifier.type,
+            identifier=item_oneclick_id,
+            start_date=today,
+            end_date=expires,
+            fulfillment_info=None,
+        )
+        return loan
+
+
+    def circulate_item(self, patron_id, item_id, return_item=False):
+        """
+        Borrow or return a catalog item.
+        :param patron_id OneClick internal id
+        :param item_id isbn
+        :return A dictionary of information on the transaction or error status and message
+            Calling methods are expected to use this dictionary to create XxxInfo objects.
+        """
+        url = "%s/libraries/%s/patrons/%s/checkouts/%s" % (self.base_url, str(self.library_id), patron_id, item_id)
+
+        method = "post"
+        action = "checkout"
+        if return_item:
+            method = "delete"
+            action = "checkin"
+
+        try:
+            response = self.request(url=url, method=method)
+        except Exception, e:
+            self.log.error("Item checkout/return failed: %r", e, exc_info=e)
+            raise RemoteInitiatedServerError(e.message)
+
+        resp_dict = {}
+        message = None
+        if response.text:
+            resp_dict = response.json()
+            message = resp_dict.get('message', None)
+
+        error_response = self.validate_response(response=response, message=message, action=action)
```

Comment:
I was confused by the fact that you assign the return value and don't use it. It's not a big deal but I would say just call `validate_response`--that makes it more clear what is going on.